Ban and kick reason length fix

ks129 opened the pull request:
Closes #861 

### Description
This PR main direction is to shorten reason in places, where is limit of characters. Main places is: Kick and Ban Audit Log entries, user notification and ModLog entries. This also limits ModLog entry description length. Added limits to watching too.

### Still to-do

- [x] Add test for characters limits
- [x] Remove DB entries when applying infraction fail.

ikuyarihS commented on bot/cogs/moderation/infractions.py:
```diff
@@ -225,7 +225,10 @@ def __init__(self, bot: Bot):
+        if len(reason) > 512:
+            log.info("Kick reason is longer than 512 characters. Reason will be truncated for Audit Log.")
+
+        action = user.kick(reason=f"{reason[:509]}..." if len(reason) > 512 else reason)
```
This can add a `...` to the middle of a word, like "See you ne..." - You will want to truncate at whitespace, `textwrap` will be of use here.

ikuyarihS commented on bot/cogs/moderation/infractions.py:
```diff
@@ -244,7 +247,14 @@ def __init__(self, bot: Bot):
+        if len(reason) > 512:
+            log.info("Ban reason is longer than 512 characters. Reason will be truncated for Audit Log.")
+
+        action = ctx.guild.ban(
+            user,
+            reason=f"{reason[:509]}..." if len(reason) > 512 else reason,
+            delete_message_days=0
+        )
```
Should we put this in a try except and move it before we call `utils.post_infraction()`? It will be rare, but there can be situation where the actual discord ban / kick failed, but since we've already put the infraction in the DB the bot will think that the ban is active.

ks129 replied:
Should this after coroutine making or after `apply_infraction`?

ks129 replied:
Also, I can't move to add into DB to end of the function: `infraction` is required for applying infraction, and in coroutine creation, this will not raise an error. An error will be raised on `self.apply_infraction`, but this need DB returned infraction.

ikuyarihS commented on bot/cogs/moderation/infractions.py:
```diff
@@ -225,7 +226,10 @@ def __init__(self, bot: Bot):
+        if len(reason) > 512:
+            log.info("Kick reason is longer than 512 characters. Reason will be truncated for Audit Log.")
+
+        action = user.kick(reason=textwrap.shorten(reason, width=509, placeholder="..."))
```
The `width` will ensure the new string will never be greater than the limit, **after appending placeholder** so you can change it to this

```suggestion
        action = user.kick(reason=textwrap.shorten(reason, width=512, placeholder=" ..."))
```

I suggest adding a space for `placeholder` as well.

You can do the same for `ban` and it'll look good to go.

MarkKoz commented on bot/cogs/moderation/infractions.py:
```diff
@@ -225,7 +226,10 @@ def __init__(self, bot: Bot):
         self.mod_log.ignore(Event.member_remove, user.id)
 
-        action = user.kick(reason=reason)
+        if len(reason) > 512:
+            log.info("Kick reason is longer than 512 characters. Reason will be truncated for Audit Log.")
```
I don't think these logs are necessary.

MarkKoz commented on bot/cogs/moderation/infractions.py:
```diff
@@ -244,7 +248,14 @@ def __init__(self, bot: Bot):
         self.mod_log.ignore(Event.member_remove, user.id)
 
-        action = ctx.guild.ban(user, reason=reason, delete_message_days=0)
+        if len(reason) > 512:
+            log.info("Ban reason is longer than 512 characters. Reason will be truncated for Audit Log.")
```
I don't think these logs are necessary.

MarkKoz commented on bot/cogs/moderation/infractions.py:
```diff
@@ -244,7 +248,14 @@ def __init__(self, bot: Bot):
+        if len(reason) > 512:
+            log.info("Ban reason is longer than 512 characters. Reason will be truncated for Audit Log.")
+
+        action = ctx.guild.ban(
+            user,
+            reason=textwrap.shorten(reason, width=512, placeholder=" ..."),
+            delete_message_days=0
+        )
```
If you put the reason in a separate variable above you could avoid having to break up the function arguments for `ban()`.

MarkKoz commented on bot/cogs/moderation/scheduler.py:
```diff
@@ -84,7 +84,8 @@ def mod_log(self) -> ModLog:
         icon = utils.INFRACTION_ICONS[infr_type][0]
-        reason = infraction["reason"]
+        # Truncate reason when it's too long to avoid raising error on sending ModLog entry
+        reason = textwrap.shorten(infraction["reason"], width=1900, placeholder="...")
```
Isn't 1900 cutting it a bit too close? That's only a bit over 100 characters for everything else.

sco1 replied:
Why is being done here? The truncation should be happening in the methods that are sending the messages, e.g. `utils.notify_infraction` and `mod_log.send_log_message`. This way you can use the actual character limit instead of having to guess at how long it needs to be in order to not raise an error.

ks129 replied:
@sco1 I applied this to `utils.notify_infraction`, but I can't do this in `mod_log.send_log_message`, because this take argument `text`, where is the full content of embed description (that include reason).

sco1 replied:
Yes, that’s the point. You’re focusing on truncating only the reason, arbitrarily guessing at the length you need to truncate to, when you can just truncate the entire string that these methods generate before sending.

ks129 replied:
@sco1 But like when reason is too long, this will cut `Expiry` (`apply_infraction`) from ModLog entry, because this is after reason. To make this, I have to put everything else first and after this reason.

sco1 replied:
Yes.

ks129 commented:
@MarkKoz Now this should be fixed.

ks129 commented:
@sco1 I now made changes, but problem is, I can't change `pardon_infraction`. This buildup and log text generating don't allow this way.

sco1 commented:
Why do you need to change `pardon_infraction`?

MarkKoz commented on bot/cogs/moderation/utils.py:
```diff
@@ -132,11 +132,11 @@
     embed = discord.Embed(
-        description=textwrap.dedent(f"""
+        description=textwrap.shorten(textwrap.dedent(f"""
             **Type:** {infr_type.capitalize()}
             **Expires:** {expires_at or "N/A"}
             **Reason:** {reason or "No reason provided."}
-            """),
+            """), width=2048, placeholder="..."),
```
This is hard to read. Split them up by assigning the dedent to a variable above.

MarkKoz commented on bot/cogs/moderation/scheduler.py:
```diff
@@ -127,7 +128,7 @@ def mod_log(self) -> ModLog:
             )
 
-            end_msg = f" (reason: {infraction['reason']})"
+            end_msg = f" (reason: {textwrap.shorten(infraction['reason'], width=1500, placeholder='...')})"
```
```suggestion
            end_msg = f" (reason: {textwrap.shorten(reason, width=1500, placeholder='...')})"
```

MarkKoz commented on bot/cogs/moderation/scheduler.py:
```diff
@@ -325,7 +326,7 @@ def mod_log(self) -> ModLog:
             "Member": f"<@{user_id}>",
             "Actor": str(self.bot.get_user(actor) or actor),
-            "Reason": infraction["reason"],
+            "Reason": textwrap.shorten(infraction["reason"], width=1500, placeholder="..."),
```
This is not ideal but I recognise this case is more troublesome than others. Do you think there is a simple way to ensure the reason will be at the end given this gets merged with another dict may have additional keys added by `pardon_infraction`? I suspect it will involve having to delete the reason from the dict then add it back in. Maybe just not worth the trouble.

ks129 replied:
Maybe should this overwrite `Reason` in `pardon_infraction` to put this as the last key? And after this, `textwrap.shorten` whole ModLog embed description?

ks129 commented:
@MarkKoz Now should OK.

MarkKoz commented on bot/cogs/watchchannels/talentpool.py:
```diff
@@ -224,7 +224,7 @@ def _nomination_to_string(self, nomination_object: dict) -> str:
                 Date: {start_date}
                 Actor: {actor.mention if actor else actor_id}
-                Reason: {nomination_object["reason"]}
+                Reason: {textwrap.shorten(nomination_object["reason"], width=200, placeholder="...")}
```
Did you confirm that these need to be shortened and that the paginator doesn't help?

ks129 replied:
Yes, because this takes it as one line.

MarkKoz commented:
All the truncation seems to be fine though.

lemonsaurus commented:
@ks129 Can you please update this pull request with a proper description? It's not acceptable to just write "Closing #861" as your pull request description - the description is supposed to provide a high level overview over specifically what this pull request does.

I'd also like to suggest that this PR address two more related and fairly minor issues:
- If the ban _does_ fail, for some other reason, it would be nice if we could reset the state properly so we don't get an erroneous state. This includes removing the record from our database and deleting the DM that was sent to the member so they don't get a false infraction notice.
- We should add a test that tests a ban with a ban reason of over 512 characters.

ks129 commented:
@lemonsaurus I fixed PR description. About tests, there are currently no tests for this moderation side.

lemonsaurus commented:
@ks129 looks like you're right, I thought we had more tests for moderation. We do have the tests in here https://github.com/python-discord/bot/blob/master/tests/bot/cogs/moderation/test_silence.py but not for most of the other moderator commands.

It would still be nice to see a single test, specifically for this problem. Would you be able to write that? No need to write a complete suite of tests for the entire moderation cog, but we got to start somewhere, right?

ks129 commented:
@lemonsaurus So I should write test only for these functions that have shortening?

lemonsaurus commented:
doesn't this just apply to bans?

ks129 commented:
@lemonsaurus I added truncation to ban, kick and ModLog entries. Should I write to test to every of them?

lemonsaurus commented:
Yeah, that seems fine to me. They can all be in the same class, though. You're basically just testing truncation.

also, we should probably get tests up and running for moderation features. That's gotta be one of the most important cogs we have. It's funny that we have extensive testing for Duck Pond but we don't have any tests for our most crucial commands.

ks129 commented:
@lemonsaurus I added now removing infraction from DB when applying fail and moved DM sending to later, so this check is this failed before sending it. Also made tests, but ban and kick reason shortening is in one file, ModLog's one is in another file.

lemonsaurus commented on tests/bot/cogs/moderation/test_infractions.py:
```diff
@@ -0,0 +1,54 @@
+from tests.helpers import MockBot, MockContext, MockGuild, MockMember, MockRole
+
+
+class ShorteningTests(unittest.IsolatedAsyncioTestCase):
```
Let's call it Truncation, not Shortening. Do you mind changing all the occurrances in this file?

lemonsaurus commented on tests/bot/cogs/moderation/test_modlog.py:
```diff
@@ -0,0 +1,29 @@
+        self.cog = ModLog(self.bot)
+        self.channel = MockTextChannel()
+
+    async def test_log_entry_description_shortening(self):
```
this one too.

thomaspet commented on bot/cogs/moderation/modlog.py:
```diff
@@ -98,7 +98,10 @@ async def send_log_message(
-        embed = discord.Embed(description=text)
+        # Truncate string directly here to avoid removing newlines
+        embed = discord.Embed(
+            description=text[:2046] + "..." if len(text) > 2048 else text
```
This will create a 2049 length string, if it's above 2048

lemonsaurus replied:
Yeah, I'm a little suspicious of these hardcoded values too. Can you justify them? Is there a test for this particular part of the code? Have you tested this manually?

ks129 replied:
I will fix this length issue, but I can't change this. Otherwise, we are getting [this issue](https://github.com/python-discord/bot/pull/864#pullrequestreview-390849141). This already tests it [here](https://github.com/ks129/bot/blob/ban-kick-reason-length/tests/bot/cogs/moderation/test_modlog.py#L17), but I have same issue there too.

lemonsaurus replied:
That's fine, just make sure the values are correct. It's mostly this length issue I was concerned about.

lemonsaurus commented on tests/bot/cogs/moderation/test_infractions.py:
```diff
@@ -0,0 +1,55 @@
+        ban = self.cog.apply_infraction.call_args[0][3]
+        self.assertEqual(
+            ban.cr_frame.f_locals["kwargs"]["reason"],
+            textwrap.shorten("foo bar" * 3000, 512, placeholder=" ...")
```
Is there supposed to be a space before the `...` here? Doesn't seem like you're doing that anywhere else. If so, why?

lemonsaurus commented on tests/bot/cogs/moderation/test_infractions.py:
```diff
@@ -0,0 +1,55 @@
+            ban.cr_frame.f_locals["kwargs"]["reason"],
+            textwrap.shorten("foo bar" * 3000, 512, placeholder=" ...")
+        )
+        # Await ban to avoid warning
+        await ban
```
Can you be more specific? What warning does this raise, and how does this avoid it? This code seems weird to me.

lemonsaurus commented on tests/bot/cogs/moderation/test_modlog.py:
```diff
@@ -0,0 +1,29 @@
+        self.channel = MockTextChannel()
+
+    async def test_log_entry_description_truncation(self):
+        """Should truncate embed description for ModLog entry."""
```
Should be written in imperial mood.

```suggestion
        """Test that embed description for ModLog entry is truncated."""
```

lemonsaurus commented on bot/cogs/moderation/infractions.py:
```diff
@@ -244,7 +245,9 @@ async def apply_ban(self, ctx: Context, user: UserSnowflake, reason: str, **kwar
         self.mod_log.ignore(Event.member_remove, user.id)
 
-        action = ctx.guild.ban(user, reason=reason, delete_message_days=0)
+        truncated_reason = textwrap.shorten(reason, width=512, placeholder=" ...")
```
Here's another placeholder with a leading space.

ks129 commented on tests/bot/cogs/moderation/test_infractions.py:
```diff
@@ -0,0 +1,55 @@
+            textwrap.shorten("foo bar" * 3000, 512, placeholder="...")
+        )
+        # Await ban to avoid not awaited coroutine warning
+        await ban
```
Addressed in 96b0261. This is a big simplification.

MarkKoz replied:
I think there's also a simple way to assert `apply_infraction` is called with the `ban` or `kick` coroutine. Since `ban` and `kick` are mocked, they'll always return the same exact mock when you call them.

ks129 replied:
Did that in 3233174.

MarkKoz replied:
Very nice.

ks129 commented on bot/cogs/moderation/scheduler.py:
```diff
@@ -164,11 +149,37 @@ async def apply_infraction(
+            dm_log_text = "\nDM: **Canceled**"
+            dm_result = f"{constants.Emojis.failmail} "
+            log.trace(f"Deleted infraction {infraction['id']} from database because applying infraction failed.")
+            await self.bot.api_client.delete(f"bot/infractions/{infraction['id']}")
```
Did that in f42ddf6, but is this enough and good?

MarkKoz replied:
Yeah, it looks fine.

ks129 commented on bot/cogs/moderation/scheduler.py:
```diff
@@ -164,11 +149,37 @@ async def apply_infraction(
         await ctx.send(
-            f"{dm_result}{confirm_msg} **{infr_type}** to {user.mention}{expiry_msg}{end_msg}."
+            f"{dm_result}{confirm_msg} "
+            f"{f'**{infr_type}** to {user.mention}{expiry_msg}{end_msg}' if not failed else ''}."
```
Addressed in 7f827ab.

ks129 commented on bot/cogs/moderation/scheduler.py:
```diff
@@ -84,6 +84,7 @@ async def apply_infraction(
         """Apply an infraction to the user, log the infraction, and optionally notify the user."""
         infr_type = infraction["type"]
         icon = utils.INFRACTION_ICONS[infr_type][0]
+        # Truncate reason when it's too long to avoid raising error on sending ModLog entry
```
Removed in 136ef11.

MarkKoz commented on bot/cogs/moderation/scheduler.py:
```diff
@@ -174,12 +174,15 @@ async def apply_infraction(
             f"{dm_result}{confirm_msg} "
-            f"{f'**{infr_type}** to {user.mention}{expiry_msg}{end_msg}' if not failed else ''}."
+            f"{infr_message}."
         )
```
Because it's now a lot shorter, it should all be on one line instead of being broken up.

ks129 replied:
Resolved in e236113.

MarkKoz replied:
The space should also be moved to the `infr_message` rather than here. Otherwise, there will be a space before the full stop when `failed == True`.

ks129 replied:
Space moved in f1f2c48.
